### ISSUE TITLE
[prefer-readonly-parameter-types] false positive with tagged primitive type

### ISSUE THREAD

**Issue by ckknight** [typescript-eslint/typescript-eslint#1790]
**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/prefer-readonly-parameter-types": "error"
  }
}
```

```TS
type MySpecialString = string & {
  readonly ' __tag': unique symbol;
};
function custom1(arg: MySpecialString) {}

// or

declare const specialTag: unique symbol;
type MyOtherSpecialString = string & {
  readonly [specialTag]: true;
}
function custom2(arg: MyOtherSpecialString) {}
```

**Expected Result**
No error

**Actual Result**
Parameter should be a read only type

**Additional Info**

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `2.25.0` |
| `@typescript-eslint/parser`        | `2.25.0` |
| `TypeScript`                       | `3.8.3` |
| `ESLint`                           | `6.8.0` |
| `node`                             | `12.16.1` |
| `npm`                              | `6.13.4` |

**Comment by bradzacher** [comment]
This case was ignored intentionally.

Primitives are treated as readonly because they're passed by value.

A intersection type will automatically be resolved at the top level by TS's type API. Which means we don't have to do any logic to determine what the resulting properties of a type will be. This logic is rather complex, so we don't want to implement it.

Unfortunately when there's a primitive in the union, we cannot outright make the assumption that "this type is readonly", because you could define perfectly valid code like this:
```js
let test: number = 1;
Object.defineProperty(String.prototype, 'test', {
  get: function () { return test },
  set: function (v) { test = v },
});

type MyString = string & { test: number }
```

However, TS does not provide an API for us to remove a single type from an intersection, which means that our only real choice for handling this would be to implement the aforementioned logic to determine the final props of the type by manually merging the non-primitive object types, which we won't be doing for the general case.

You can use the `Readonly` type in this case to make the primitive intersection type readonly, and the rule will recognise it as such.

----

That being said, I am happy to accept a PR which handles the common case of a primitive and a single object, like your example.
By only handling a single item in the union, there is no merging logic required, so it should be pretty easy.

**Comment by ScrapsPrograms** [comment]
I found another interesting case, potentially another false positive, if I haven't made any mistakes in my code.
Here's the snippet, taken from my NestJS project:

```ts
const BodyParams: (
  ...pipes: readonly (Readonly<PipeTransform> | Readonly<Type<PipeTransform>>)[]
) => // Implementation
```

The error shows on the '...pipes' rest parameter, which evidently is fully typed as readonly. This goes for both the array as well as all the elements within the array.
